### PR TITLE
API Gateway private REST API - Support VPC Endpoint

### LINKED ISSUE #10154: Amazon API Gateway private REST APIs with VPC endpoints
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Amazon API Gateway simplifies accessing private APIs by allowing you to associate one or more Amazon Virtual Private Cloud (VPC) Endpoints to a private API.

### New or Affected Resource(s)

* [`api_gateway_rest_api`](https://www.terraform.io/docs/providers/aws/r/api_gateway_rest_api.html)

### Potential Terraform Configuration

```hcl
resource "aws_api_gateway_rest_api" "example" {
  name = "example"

  endpoint_configuration {
    types = ["PRIVATE"]
    vpc_endpoint_ids = ["${aws_vpc_endpoint.example.id}"]
  }
}
```

### References

[Announcement](https://aws.amazon.com/about-aws/whats-new/2019/09/amazon-api-gateway-simplifies-invoking-private-apis/).
[Developer guide](https://docs.aws.amazon.com/apigateway/latest/developerguide/associate-private-api-with-vpc-endpoint.html).

Requires [AWS SDK v1.24.1](https://github.com/aws/aws-sdk-go/releases/tag/v1.24.1):
* https://github.com/terraform-providers/terraform-provider-aws/pull/10011

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Closes #10154

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource/aws_api_gateway_rest_api: Add vpc_endpoint_ids argument
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSAPIGatewayRestApi_'

--- PASS: TestAccAWSAPIGatewayRestApi_basic (138.81s)
--- PASS: TestAccAWSAPIGatewayRestApi_EndpointConfiguration (121.38s)
--- PASS: TestAccAWSAPIGatewayRestApi_EndpointConfiguration_Private (151.55s)
--- PASS: TestAccAWSAPIGatewayRestApi_EndpointConfiguration_VPCEndpoint (506.39s)
--- PASS: TestAccAWSAPIGatewayRestApi_api_key_source (431.64s)
--- PASS: TestAccAWSAPIGatewayRestApi_policy (107.07s)
--- PASS: TestAccAWSAPIGatewayRestApi_openapi (75.58s)
...
```

### CODE REVIEW

File: aws/resource_aws_api_gateway_rest_api.go [hunk 1]
```diff
@@ -335,6 +343,30 @@ func resourceAwsApiGatewayRestApiUpdateOperations(d *schema.ResourceData) []*api
...
+				Op:   aws.String("remove"),
+				Path: aws.String(fmt.Sprintf("%s/%s", prefix, escapeJsonPointer(v.(string)))),
+			})
+		}
+
+		if len(new) > 0 {
+			for _, v := range new {
+				operations = append(operations, &apigateway.PatchOperation{
+					Op:   aws.String("add"),
+					Path: aws.String(fmt.Sprintf("%s/%s", prefix, escapeJsonPointer(v.(string)))),
```
Comment (folbricht) [inline aws/resource_aws_api_gateway_rest_api.go]: This doesn't appear to work for a few reasons:
- The path should be `/endpointConfiguration/vpcEndpointIds` according to https://docs.aws.amazon.com/apigateway/latest/developerguide/associate-private-api-with-vpc-endpoint.html
- Should use `endpoint_configuration.0.` not `.1`

Here's an alternative implementation. Tested in a real deployment.

```go
	if d.HasChange("endpoint_configuration.0.vpc_endpoint_ids") {
		o, n := d.GetChange("endpoint_configuration.0.vpc_endpoint_ids")
		prefix := "/endpointConfiguration/vpcEndpointIds"

		old := o.([]interface{})
		new := n.([]interface{})

		for _, v := range old {
			operations = append(operations, &apigateway.PatchOperation{
				Op:    aws.String("remove"),
				Path:  aws.String(prefix),
				Value: aws.String(v.(string)),
			})
		}

		for _, v := range new {
			operations = append(operations, &apigateway.PatchOperation{
				Op:    aws.String("add"),
				Path:  aws.String(prefix),
				Value: aws.String(v.(string)),
			})
		}
	}
```

File: aws/resource_aws_api_gateway_rest_api.go [hunk 2]
```diff
@@ -410,7 +446,8 @@ func flattenApiGatewayEndpointConfiguration(endpointConfiguration *apigateway.En
 	}
 
 	m := map[string]interface{}{
-		"types": flattenStringList(endpointConfiguration.Types),
+		"types":            flattenStringList(endpointConfiguration.Types),
+		"vpc_endpoint_ids": flattenStringList(endpointConfiguration.VpcEndpointIds),
```
Comment (folbricht) [inline aws/resource_aws_api_gateway_rest_api.go]: This causes failures with other endpoints not of type `PRIVATE` that don't have VPC endpoint IDs. The failures look like this:
```
Error: error setting endpoint_configuration: Invalid address to set: []string{"endpoint_configuration", "0", "vpc_endpoint_ids"}

  on ../modules/api_gateway_domain/main.tf line 5, in resource "aws_api_gateway_domain_name" "api_gateway_domain":
   5: resource "aws_api_gateway_domain_name" "api_gateway_domain" {
```

Tested alternative implementation:
```go
	m := map[string]interface{}{
		"types": flattenStringList(endpointConfiguration.Types),
	}

	if len(endpointConfiguration.VpcEndpointIds) > 0 {
		m["vpc_endpoint_ids"] = flattenStringList(endpointConfiguration.VpcEndpointIds)
	}
```

Comment (DrFaust92) [reply aws/resource_aws_api_gateway_rest_api.go]: i missed that this function also affects the domain resource

File: aws/resource_aws_api_gateway_rest_api.go (hunk 1 above)
Comment (DrFaust92) [reply aws/resource_aws_api_gateway_rest_api.go]: fixed, thanks!

File: aws/resource_aws_api_gateway_rest_api.go (hunk 2 above)
Comment (DrFaust92) [reply aws/resource_aws_api_gateway_rest_api.go]: fixed!

**Comment by DrFaust92** [comment]
fixed according to @folbricht comments:

```
$ make testacc TESTARGS='-run=TestAccAWSAPIGatewayRestApi_'

=== RUN   TestAccAWSAPIGatewayRestApi_basic
=== PAUSE TestAccAWSAPIGatewayRestApi_basic
=== CONT  TestAccAWSAPIGatewayRestApi_basic
--- PASS: TestAccAWSAPIGatewayRestApi_basic (118.65s)
=== RUN   TestAccAWSAPIGatewayRestApi_EndpointConfiguration
=== PAUSE TestAccAWSAPIGatewayRestApi_EndpointConfiguration
=== CONT  TestAccAWSAPIGatewayRestApi_EndpointConfiguration
--- PASS: TestAccAWSAPIGatewayRestApi_EndpointConfiguration (127.70s)
=== RUN   TestAccAWSAPIGatewayRestApi_EndpointConfiguration_Private
=== PAUSE TestAccAWSAPIGatewayRestApi_EndpointConfiguration_Private
=== CONT  TestAccAWSAPIGatewayRestApi_EndpointConfiguration_Private
--- PASS: TestAccAWSAPIGatewayRestApi_EndpointConfiguration_Private (121.22s)
=== RUN   TestAccAWSAPIGatewayRestApi_EndpointConfiguration_VPCEndpoint
=== PAUSE TestAccAWSAPIGatewayRestApi_EndpointConfiguration_VPCEndpoint
=== CONT  TestAccAWSAPIGatewayRestApi_EndpointConfiguration_VPCEndpoint
--- PASS: TestAccAWSAPIGatewayRestApi_EndpointConfiguration_VPCEndpoint (574.07s)
=== RUN   TestAccAWSAPIGatewayRestApi_api_key_source
=== PAUSE TestAccAWSAPIGatewayRestApi_api_key_source
=== CONT  TestAccAWSAPIGatewayRestApi_api_key_source
--- PASS: TestAccAWSAPIGatewayRestApi_api_key_source (335.67s)
=== RUN   TestAccAWSAPIGatewayRestApi_policy
=== PAUSE TestAccAWSAPIGatewayRestApi_policy
=== CONT  TestAccAWSAPIGatewayRestApi_policy
--- PASS: TestAccAWSAPIGatewayRestApi_policy (149.33s)
=== RUN   TestAccAWSAPIGatewayRestApi_openapi
=== PAUSE TestAccAWSAPIGatewayRestApi_openapi
=== CONT  TestAccAWSAPIGatewayRestApi_openapi
--- PASS: TestAccAWSAPIGatewayRestApi_openapi (88.30s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	670.232s

```

**Comment by davidmco65** [comment]
Any idea when this will be merged?  I've literally got this exact problem holding my deployment up right now.  Thanks!

**Comment by cleverton-heusner** [comment]
Hi, guys.

Any news about this merge request? We really need it in our project. Thanks a lot!

**Comment by DrFaust92** [comment]
@bflad, can you help out with this? 

For some reason it I can restore the review request on the PR

**Comment by dininski** [comment]
Any news on merging this PR? We would also benefit massively from this feature in our project.

Thanks in advance and all the best!

**Comment by UsmanMohammad** [comment]
Hi,

We are also looking for this feature as well.

Many thanks

**Comment by gncj** [comment]
Hello,

any update on this request?

**Comment by tokiwong** [comment]
Hello,

I am also looking for this feature

**Comment by skyguy94** [comment]
I'm looking for this feature as well.

**Comment by cleblanc87** [comment]
Any feedback from the core team on this PR?

**Comment by aaron0browne** [comment]
In case it's useful to anyone here, this is my workaround for the moment, a null resource provisioner to run the aws cli command that associates a vpc endpoint with the api gateway. This obviously assumes you have a the referenced resources elsewhere in your config.

```
# Associate the API Gateway with the VPC endpoint so that we can access it without using a custom header,
#   which is not possible in browser pre-flight CORS requests.
# See: https://forums.aws.amazon.com/thread.jspa?messageID=916693
# And: https://docs.aws.amazon.com/apigateway/latest/developerguide/associate-private-api-with-vpc-endpoint.html
# Terraform PR to add this ability here: https://github.com/terraform-providers/terraform-provider-aws/pull/10627
resource null_resource api_vpc_endpoint {
  provisioner local-exec {
    command = join("", [
      "aws apigateway update-rest-api ",
      "--rest-api-id ${aws_api_gateway_rest_api.default.id} ",
      "--patch-operations \"op='add',path='/endpointConfiguration/vpcEndpointIds',value='${data.aws_vpc_endpoint.api_gateway.id}'\" ",
      "--region ${var.region}",
    ])
  }
}
```

**Comment by DrFaust92** [comment]
I'd like to add `terraform-providers/aws-provider` as a reviewer but am not able to for some reason.
Can anyone from the core team assist? @bflad @aeschright

**Comment by aaron0browne** [comment]
Is anyone else experience unreliability with this configuration? I am seeing a successful `apply`, no errors or failures, but the api gateway is still NOT associated with the VPCE.

**Pull request merged**